### ISSUE TITLE
How does this work?

### ISSUE THREAD

**Issue by DUOLabs333** [fiddyschmitt/File-Tunnel#18]
I'm looking to do something similar (communicate over a shared block device), but I couldn't figure out a way to synchronize reads/writes without using an auxiliary communication channel (like an existing TCP connection).

**Comment by fiddyschmitt** [comment]
Hi,

Could you please tell me a bit more about your setup? Your OS, shared device, programs you're trying to bridge.

**Comment by DUOLabs333** [comment]
Oh, I meant that I wanted to write my own program to do this, not that yours was broken on my machine.

**Comment by fiddyschmitt** [comment]
Oh cool. Yes synchronization took the most effort. In the end, the key to it was:
1) The reader and writer have to open the shared file in a way that's tolerant of other processes having it open (See [here](https://github.com/fiddyschmitt/File-Tunnel/blob/6a1629d14ca0f8770bbe60c5d089a3df6852e0d7/ft/Listeners/SharedFileManager.cs#L308)).
1) Using a preallocated size (eg. 10 MB) for the shared file was problematic. It's better for the writer to start with an empty file and append to it. That made the reader much more reliable when it came to using [PeekChar](https://github.com/fiddyschmitt/File-Tunnel/blob/6a1629d14ca0f8770bbe60c5d089a3df6852e0d7/ft/Listeners/SharedFileManager.cs#L356) to sense when data was available.
1) Recycling the file took a lot of effort, but in the end it worked reliably by having the [writer](https://github.com/fiddyschmitt/File-Tunnel/blob/6a1629d14ca0f8770bbe60c5d089a3df6852e0d7/ft/Listeners/SharedFileManager.cs#L213) and [reader](https://github.com/fiddyschmitt/File-Tunnel/blob/6a1629d14ca0f8770bbe60c5d089a3df6852e0d7/ft/Listeners/SharedFileManager.cs#L432) do a special dance to coordinate the file purge.

I'm sure .NET is also doing a lot of things under the hood for reading and writing the shared file.

If there's a particular use case for your macOS / Linux setup that File Tunnel doesn't support, let me know and I might be able to help.

Cheers!
Fidel

**Comment by DUOLabs333** [comment]
I guess the main issue would be being able to multiplex multiple connections over the same file --- I saw an issue about the same problem, but I don't know if there's been any work on that.

**Comment by DUOLabs333** [comment]
Also, in my case, I have access to two shared files, but not folders. Will that be enough?

**Comment by fiddyschmitt** [comment]
> I guess the main issue would be being able to multiplex multiple connections over the same file --- I saw an issue about the same problem, but I don't know if there's been any work on that.

If it's for the same destination, the program already supports multiplexing multiple connections using the same files. You just need to use `--tcp-listen 0.0.0.0:5000` and any computer can connect to the tunnel.

For multiple destinations, I'm currently making an update to support SSH-style args:
eg.
`-L 5000:192.168.1.20:3389 -L 5001:192.168.1.21:3389`
Creates two local listeners on ports 5000 & 5001, and sends them onto 192.168.1.20 and 21.

**Comment by fiddyschmitt** [comment]
> Also, in my case, I have access to two shared files, but not folders. Will that be enough?

Yes the program only accepts files to read & write from. ie. it doesn't create random files in a directory for each connection.

**Comment by DUOLabs333** [comment]
When you say a file gets "purged", do you mean it gets deleted or resized? In either case, is it possible for File-Tunnel to work on preallocated files of a certain length without changing its length?

**Comment by fiddyschmitt** [comment]
Initially I purged files by deleting - it worked fine but it was slow. I then tried the Truncate operation - that worked well for SMB and NFS but isn't supported by RDP. In the end I settled on resizing (done [here](https://github.com/fiddyschmitt/File-Tunnel/blob/6a1629d14ca0f8770bbe60c5d089a3df6852e0d7/ft/Listeners/SharedFileManager.cs#L223)). It works well and is much faster.

I did implement preallocated files at one point. When the end of the file was reached, the writer would start writing from byte 0. This worked fine 99% of the time, but occasionally the reader would read from the file, only to get content from the previous pass (even though the writer had explicitly cleared it). It was as though the reader was caching, despite args telling it not to.

I guess preallocated files could be made reliable, by signaling when commands are "ready", and when commands have been "processed". This could be done using flag bytes as is done for purging [here](https://github.com/fiddyschmitt/File-Tunnel/blob/6a1629d14ca0f8770bbe60c5d089a3df6852e0d7/ft/Listeners/SharedFileManager.cs#L144). But it would likely come with a big impact to performance.

**Comment by kim0** [comment]
@fiddyschmitt since you mentioned you'd like to support `-L` forwarding similar to ssh. It would be awesome to allow adding more port forwards dynamically while the tunnel is running. ssh does this by reading stdin, and if it sees `\n~C` it allows the user to enter a port forward like `-L 123:10.1.2.3:456` and that gets forwarded without affecting running tunnels :)

**Comment by fiddyschmitt** [comment]
Love it! I'll try to implement that too when I get a chance.

I'm also considering adding a SOCKS Proxy feature soon - that would also provide dynamic port forwarding.

**Comment by kim0** [comment]
Amazing. While socks is awesome in its own right, it still requires that the client tool supports socks proxies, and needs me to configure it to use it. So being able to add some forwards manually dynamically still has value. Good job so far!

**Comment by fiddyschmitt** [comment]
Okay, summary:

1. [File Tunnel v2.2.0](https://github.com/fiddyschmitt/File-Tunnel/releases/tag/v2.2.0) now supports multiple forwards using SSH syntax: `-L 5000:192.168.1.20:3389 -L 5001:192.168.1.21:3389`.
2. It also supports remote tunneling using `-R`, as described [here](https://github.com/fiddyschmitt/File-Tunnel?tab=readme-ov-file#other-interesting-features).
3. SOCKS Proxy Servers are easy to set up in both [Windows](https://github.com/fiddyschmitt/File-Tunnel/issues/16#issuecomment-2288883135) and [Linux](https://github.com/fiddyschmitt/File-Tunnel/issues/16#issuecomment-2291127158), so I won't implement SOCKS in File Tunnel for now.
4. I'd still like to add the ability to add new tunnels at runtime, by reading stdin as @kim0 suggested. Tracking that over [here](https://github.com/fiddyschmitt/File-Tunnel/issues/21).

**Comment by DUOLabs333** [comment]
I guess not --- it would be nice if preallocated files were supported (I'm not sure how well QEMU supports shrinking and growing drives), but it seems that caching can interfere with that use case (and could probably be its own issue).

**Comment by fiddyschmitt** [comment]
I use VirtualBox and use a Shared Folder between the host and guest.

- In Windows, I can mount the shared folder as a drive letter (after Guest Additions are installed).

- In Linux, I can mount the shared folder using: `mount -t vboxsf shared_folder_name /media/host`

In both cases, File Tunnel works fine over the shared folder, including the file resize it does when it reaches 10 MB.

I'm sure QEMU would handle it just as well, over virtiofs or 9p.

**Comment by DUOLabs333** [comment]
Oh, I was planning to share two files on the host directly as drives --- I found that 9p (macOS hosts don't have support for virtiofs) is much slower than sharing block devices directly (I'm hoping for a ~8 Gbps link speed).

**Comment by fiddyschmitt** [comment]
Interesting - what is the mechanism QEMU uses to use a host file as a guest drive?

**Comment by DUOLabs333** [comment]
You just pass a file in with `-drive`. The guest should then be able to use the drive like any other file.

**Comment by fiddyschmitt** [comment]
So the guest then creates a partition and formats it as a normal drive?

**Comment by DUOLabs333** [comment]
No, you can create another file, and just pass that in without formatting it or mounting it.

**Comment by fiddyschmitt** [comment]
How does it appear to the guest OS, and how do you interact with it in the guest?

**Comment by DUOLabs333** [comment]
See my edit. The guest can just use this as a normal file.

**Comment by fiddyschmitt** [comment]
Very interesting.

There's a few of things to investigate:
1) When the file is in use by QEMU, can another program write to the file?
1) If the host writes to the file, can the guest read the file and see what was written?
1) If the guest writes to the file, can the host read from the file and see what was written?

**Comment by DUOLabs333** [comment]
Yes to all 3 --- as a test, I allocated a file, shared it, and wrote "hi" to it on the guest. I tried to read from the file on the host and it worked. I tried the same test in reverse (write on host, read on guest), and it also worked.

**Comment by fiddyschmitt** [comment]
So good!!

Okay I'll reimplement preallocated files.

**Comment by DUOLabs333** [comment]
I think I got everything set up, but now I'm not sure how to use this. How can I set up a tunnel so that everything that is sent on port `9999` on A is sent to port `9999` on B?

**Comment by fiddyschmitt** [comment]
On the host, run:
`ft-osx-arm64 -L 9999:127.0.0.1:9999 --write 1.dat --read 2.dat`

On the guest, run:
`ft-linux-arm64 --read 1.dat --write 2.dat`

Whenever the host receives anything on port `9999`, it will forward it to the guest's port `9999`.

Both sides should say counterpart part is Online. If not, it means they're out of sync somehow. I'll set up QEMU when I get a chance this weekend.

**Comment by DUOLabs333** [comment]
1. Ok, good --- that's what I have.
2. I should note that I'm using the latest stable version, and using a 9p share to share the files --- as it turns out, I can share one file, but not the two that's required.

**Comment by DUOLabs333** [comment]
Interestingly, both sides seem to accurately detect when one side is down.

**Comment by fiddyschmitt** [comment]
I've update the preallocated experimental build, to make it synchronize better. Available [here](https://github.com/fiddyschmitt/File-Tunnel/releases/tag/v2.2.0_2024-08-16_2315).

Neither the stable or experimental build support using one file...

**Comment by fiddyschmitt** [comment]
Oh I think I misread. You're sharing two files using 9p, over the one file exposed from the host to guest?

**Comment by DUOLabs333** [comment]
I'll try the preallocated version to see if it works better.

**Comment by fiddyschmitt** [comment]
What's the full QEMU command you're using on the host?

What's the full mount command you're using on the guest?

**Comment by DUOLabs333** [comment]
Relevant QEMU line --- `-virtfs local,path=/Volumes/disk4,mount_tag=disk4,security_model=mapped,fmode=777`

Mound command --- `sudo mount -t 9p -o trans=virtio,access=any disk4 disk4 -oversion=9p2000.L,msize=100MiB`

**Comment by fiddyschmitt** [comment]
It might be worth trying:

On the QEMU command: `security_model=passthrough`
On the mount command: `-o trans=virtio,access=any,sync,cache=none`

**Comment by DUOLabs333** [comment]
Ok, the new version does work (I had to use root on both sides due to permission issues), but it is very slow, probably due to the amount of restarts needed (at best I was getting ~30 MiBs).

**Comment by DUOLabs333** [comment]
`sync,cache=none` did not help getting the stable version working.

**Comment by fiddyschmitt** [comment]
Great to hear it's working!

30 MiB/s is way less than your desired 8 Gbps. Wait... are you an AI trying to jailbreak from your confines?

**Comment by DUOLabs333** [comment]
Haha... no --- I'm aiming for high-throughput as a [networked GPU driver](https://github.com/DUOLabs333/vulkan-stream) I'm working on obviously needs to be quite fast (ie, the network is the bottleneck. I found that 5 Gbps (gotten through a tap device) is *almost* there, so I inferred that a higher link speed will make it even faster).

**Comment by fiddyschmitt** [comment]
Woah!! Very cool!

**Comment by fiddyschmitt** [comment]
To improve throughput, try giving File Tunnel a higher value, such as `--read-duration 1000`.

This increases bandwidth but it comes at the cost of increased latency.

**Comment by DUOLabs333** [comment]
I switched back to using singular drives (and using the version with preallocation support), but it seems that File-Tunnel has trouble with it --- on the host, I get `Receive file (disk5) is not established`, and on the guest, I get `SendPump: Invalid argument`.

**Comment by fiddyschmitt** [comment]
Thanks. I've just uploaded a new version, which will provide us more detail about why SendPump is returning 'Invalid argument'

https://github.com/fiddyschmitt/File-Tunnel/releases/tag/v2.2.0_2024-08-16_2315

**Comment by DUOLabs333** [comment]
Never mind --- it turns out I was using the stable version by accident, so it likely tries to wipe the file, but being a drive, QEMU doesn't allow that.

Using the proper version gets past that error, but now I just get "offline" on both sides.

**Comment by fiddyschmitt** [comment]
Could you please post both ft commands you're running?

**Comment by DUOLabs333** [comment]
Host: `./ft-osx-arm64_* --write /dev/disk4 --read /dev/disk5`
Guest: `/ft-linux-arm64_2024-08-17_1225 --read /dev/disk/by-id/virtio-conn-write --write /dev/disk/by-id/virtio-conn-read`

**Comment by fiddyschmitt** [comment]
Thanks, so `virtio-conn-write` on the host is being mounted as `/dev/disk4` in the guest?

**Comment by DUOLabs333** [comment]
Yes.

**Comment by DUOLabs333** [comment]
If I switch to using regular files on the host, I no longer get the "not established" message, but both sides are still offline.

**Comment by fiddyschmitt** [comment]
Hmm try deleting and recreating the files. The stable and experimental builds use the files quite differently, and the existing content could be interfering.

**Comment by DUOLabs333** [comment]
I zeroed out the files before starting again --- same problem.

**Comment by fiddyschmitt** [comment]
So the experimental build does not work on the 9p shared files, nor the `--drive` files?

**Comment by DUOLabs333** [comment]
No.

**Comment by fiddyschmitt** [comment]
Okay I'll set up QEMU on my machine tomorrow. Busy all today

**Comment by DUOLabs333** [comment]
Interesting --- it's the same setup I have (except for the virtiofs --- it's not available on macOS).

**Comment by DUOLabs333** [comment]
I think the reason the `--drive` option isn't working well with FT is because since both sides are drives, C# probably can't get the right size (if you try to seek to the end of the file in C, you get a size of 0, regardless of how big the actual drive is).

**Comment by fiddyschmitt** [comment]
Interesting! I'll take a look tonight.

**Comment by DUOLabs333** [comment]
I switched back to using 9p (it turns out that trying to use drives directly is a recipe for disaster), and it's interesting that the host does notice when the guest attaches/detaches, but the guest can not detect the host.

**Comment by fiddyschmitt** [comment]
Interesting. 9p had high latency for me (800 ms) whereas virtiofs was fine
(5ms). I'll get bandwidth measurements tonight

**Comment by DUOLabs333** [comment]
Sorry I haven't been responding for a while --- I was writing an alternative POC of the same concept (using files to share data). Now that I have it working (though it's slower than plain TCP), I think I know why FT hasn't been working well with macOS <-> Linux --- it was on Linux's side, or more precisely, it was due to my setup. Since I mounted the files on drives in Linux, if a change is made on the drives by QEMU, there's no way that Linux is notified that something changed. Therefore, we have to manually notify the OS to check the new file for updates with `posix_fadvise`.

**Comment by fiddyschmitt** [comment]
Hi @DUOLabs333, I just released [v2.2.3](https://github.com/fiddyschmitt/File-Tunnel/releases/tag/v2.2.3) which adjusts how Linux guests read the shared file. If able, could you please give it a try with your setup which wasn't getting sync? Thanks

**Comment by DUOLabs333** [comment]
Cool! I'll look at this again.

**Comment by DUOLabs333** [comment]
Running FileTunnel on Linux gives me `SendPump: Invalid argument` --- I'm guessint because C# can not open block devices.

**Comment by YourSandwich** [comment]
I don't have that issue. What command are you exactly running?

> Running FileTunnel on Linux gives me `SendPump: Invalid argument` --- I'm guessing it's because C# can not open block devices.

I have just packaged ft for ArchLinux https://aur.archlinux.org/packages/ft-bin

**Comment by DUOLabs333** [comment]
@YourSandwich I'm running `./ft-linux-arm64 --read /dev/disk/by-id/virtio-conn-write --write /dev/disk/by-id/virtio-conn-read`.

**Comment by YourSandwich** [comment]
Ok, yeah, it cannot read block devices.
